src/group_theory/divisible.lean

```diff
@@ -0,0 +1,299 @@
+-/
+class divisible_by :=
+(div : A → α → A)
+(div_zero : ∀ a, div a 0 = 0)
+(div_cancel : ∀ {n : α} (a : A), n ≠ 0 → n • (div a n) = a)
+
+section pi
+
+variables {ι β : Type*} (B : ι → Type*) [Π (i : ι), has_scalar β (B i)]
+variables [has_zero β] [Π (i : ι), add_monoid (B i)] [Π i, divisible_by (B i) β]
+
+instance divisible_by_pi : divisible_by (Π i, B i) β :=
+{ div := λ x n i, divisible_by.div (x i) n,
+  div_zero := λ x, funext $ λ i, divisible_by.div_zero _,
+  div_cancel := λ n x hn, funext $ λ i, divisible_by.div_cancel _ hn }
+
+end pi
+
+section prod
+
+variables {β B B' : Type*} [has_zero β] [add_monoid B] [add_monoid B']
+variables [has_scalar β B] [has_scalar β B'] [divisible_by B β] [divisible_by B' β]
+
+instance divisible_by_prod : divisible_by (B × B') β :=
+{ div := λ p n, (divisible_by.div p.1 n, divisible_by.div p.2 n),
+  div_zero := λ p, prod.ext (divisible_by.div_zero _) (divisible_by.div_zero _),
+  div_cancel := λ n p hn, prod.ext (divisible_by.div_cancel _ hn) (divisible_by.div_cancel _ hn) }
+
+end prod
+
+end add_monoid
+
+namespace monoid
+
+variables (A α : Type*) [monoid A] [has_pow A α] [has_zero α]
+
+/--
+A `monoid A` is `α`-rootable iff `xⁿ = a` has a solution for all `n ≠ 0 ∈ α` and `a ∈ A`.
+Here we adopt a constructive approach where we ask an explicit `root : A → α → A` function such that
+* `root a 0 = 1` for all `a ∈ A`
+* `(root a n)ⁿ = a` for all `n ≠ 0 ∈ α` and `a ∈ A`.
+-/
+@[to_additive add_monoid.divisible_by
+"An `add_monoid A` is `α`-divisible iff `n • x = a` has a solution for all `n ≠ 0 ∈ α` and `a ∈ A`.
+Here we adopt a constructive approach where we ask an explicit `div : A → α → A` function such that
+* `div a 0 = 0` for all `a ∈ A`
+* `n • div a n = a` for all `n ≠ 0 ∈ α` and `a ∈ A`."]
+class rootable_by :=
+(root : A → α → A)
+(root_zero : ∀ a, root a 0 = 1)
+(root_cancel : ∀ {n : α} (a : A), n ≠ 0 → (root a n)^n = a)
+
+@[to_additive smul_surj_of_divisible_by]
+lemma pow_left_surj_of_rootable_by [rootable_by A α] {n : α} (hn : n ≠ 0) :
+  function.surjective (λ a, pow a n : A → A) :=
+λ x, ⟨rootable_by.root x n, rootable_by.root_cancel _ hn⟩
+
+/--
+A `monoid A` is `α`-rootable iff the `pow _ n` function is surjective, i.e. the constructive version
+implies the textbook approach.
+-/
+@[to_additive add_monoid.divisible_by_of_div_surj
+"An `add_monoid A` is `α`-divisible iff `n • _` is a surjective function, i.e. the constructive
+version implies the textbook approach."]
+noncomputable def rootable_by_of_pow_surj
+  (H : ∀ {n : α}, n ≠ 0 → function.surjective (λ a, a^n : A → A)) :
+rootable_by A α :=
+{ root := λ a n, @dite _ (n = 0) (classical.dec _) (λ _, (1 : A)) (λ hn, (H hn a).some),
+  root_zero := λ _, by classical; exact dif_pos rfl,
+  root_cancel := λ n a hn, by rw dif_neg hn; exact (H hn a).some_spec }
+
+section pi
+
+variables {ι β : Type*} (B : ι → Type*) [Π (i : ι), has_pow (B i) β]
+
+instance has_pow_pi : has_pow (Π i, B i) β :=
```

Comment:
I think this should move to `data.pi.algebra`, and then be used in the monoid `algebra.group.pi`.

Reply:
I added `has_pow_pi` and `has_pow_prod` instances, but I think I messed up the `to_additive` part, consequently, I had to change a lot of lemmas in `continuous_map` file.

Reply:
I broke `topology/continuous_function/algebra.lean`, but don't know how to fix it

Reply:
Can you do this in a standalone PR? That doesn't surprise me too much, but I ought to be an easy fix

Reply:
This one #15478